blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx

```diff
@@ -15,71 +17,140 @@ import {
+                        // we always want to refresh if the run has finished
+                        fetchNodes({ ...this.mergedConfig, refetch });
+                        break;
+                    }
+                default:
+                    {
+                        // //console.log(event);
+                    }
+                }
+            } catch (e) {
+                // we only ignore the exit error
+                if (e.message !== 'exit') {
+                    throw e;
                 }
-                break;
-            }
-            case 'pipeline_end': {
-                fetchNodes({ ...mergedConfig, refetch });
-                break;
             }
-            default: {
-                // console.log(event);
+        };
+
+        this.pipelineListener = sse.subscribe('pipeline', onSseEvent);
+    }
+
+    componentDidMount() {
+        // need to register handler to step out of karaoke mode
+        // we bail out on scroll up
+        const onScrollHandler = (elem) => {
+            if (elem.deltaY < 0 && this.state.followAlong) {
+                this.setState({ followAlong: false });
             }
+        };
+        // we bail out on arrow_up key
+        const _handleKeys = (event) => {
+            if (event.keyCode === 38 && this.state.followAlong) {
+                this.setState({ followAlong: false });
             }
-        });
+        };
+        // determine scroll area
+        const domNode = ReactDOM.findDOMNode(this.refs.scrollArea);
+        // add both listemer, one to the scroll area and another to the whole document
+        domNode.addEventListener('wheel', onScrollHandler, false);
+        document.addEventListener('keydown', _handleKeys, false);
     }
 
     componentWillReceiveProps(nextProps) {
-        if (nextProps.params.node !== this.props.params.node) {
-            const config = this.generateConfig(nextProps);
-            this.props.setNode(config);
-            this.props.fetchSteps(config);
-        }
+        const followAlong = this.state.followAlong;
+        this.mergedConfig = this.generateConfig({ ...nextProps, followAlong });
 
+        // we do not want any timeouts if we are not doing karaoke
+        if (!this.state.followAlong && this.timeout) {
+            clearTimeout(this.timeout);
+        }
+        // calculate if we need to trigger any actions to get into the right state (is plain js for testing reasons)
+        const nodeAction = calculateNode(this.props, nextProps, this.mergedConfig);
+        if (nodeAction && nodeAction.action) {
+            // use updated config
+            this.mergedConfig = nodeAction.config;
+            // we may need to stop following
+            if (this.state.followAlong !== nodeAction.state.followAlong) {
+                this.setState({ followAlong: nodeAction.state.followAlong });
+            }
+            // if we have actions we fire them
+            this.props[nodeAction.action](this.mergedConfig);
```

Comment:
This dynamic invoke might be a little hard for some people to follow. I am not strongly against it, but I wonder if there's a way to do this in a way that's a little more obvious.